repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +88,48 @@ class Session(interpreter: Interpreter)
     synchronized {
       _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
     }
+
     Future {
       synchronized {
-        _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+        // Statement may be removed by cancel operation before running.
+        if (_statements.contains(statementId)) {
+          _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+        }
       }
 
-      val statement =
+      // Statement may be removed before really executing the code, we should not execute the
+      // code if it has already been removed.
+      val stmtExisted = synchronized { _statements.contains(statementId) }
+      val statement = if (stmtExisted) {
+        _sc.foreach(_.setJobGroup(statementId.toString, s"Job group for statement ${statementId}"))
         new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      } else {
+        null
+      }
 
       synchronized {
-        _statements(statementId) = statement
+        // Check whether statement has been removed by cancel operation
+        if (_statements.contains(statementId) && statement != null) {
+          _statements(statementId) = statement
+        }
       }
     }
     statementId
   }
 
+  def cancel(statementId: Int): Unit = {
+    val stmt = synchronized {
+      _statements.remove(statementId)
```

Comment:
This will cancel executed statement.
Also it leaves holes in statement history. Instead of removing the statement, I think we should change its state to "cancelled". This's more explicit to users that a statement was submitted then cancelled.

Also, we should be careful of race conditions like cancelling a statement being executing but is succeeding. To avoid locking the whole statement map `statements`, I think we could change `Statement` to a mutable object, and when we change its state, we lock it then call its setter:
```
_statements(statementId).synchronized { 
  _statements(statementId).setState(...) 
}
```
instead of 
```_statements(statementId) = new Statement(statementId, ..., null)```.